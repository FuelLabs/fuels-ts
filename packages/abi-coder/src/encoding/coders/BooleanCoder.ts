import { ErrorCode, FuelError } from '@fuel-ts/errors';
import { bn, toBytes } from '@fuel-ts/math';

import type { EncodingOptions } from '../../types/EncodingOptions';
import { WORD_SIZE } from '../../utils/constants';

import { Coder } from './AbstractCoder';

export class BooleanCoder extends Coder<boolean, boolean> {
  options: EncodingOptions;

  constructor(
    options: EncodingOptions = {
      padToWordSize: false,
    }
  ) {
    const encodedLength = options.padToWordSize ? WORD_SIZE : 1;
    super('boolean', 'boolean', encodedLength);

    this.options = options;
  }

  encode(value: boolean): Uint8Array {
    const isTrueBool = value === true || value === false;

    if (!isTrueBool) {
      throw new FuelError(ErrorCode.ENCODE_ERROR, `Invalid boolean value.`);
    }

    return toBytes(value ? 1 : 0, this.encodedLength);
  }

  decode(data: Uint8Array, offset: number): [boolean, number] {
    if (data.length < this.encodedLength) {
      throw new FuelError(ErrorCode.DECODE_ERROR, `Invalid boolean data size.`);
    }

    const bytes = bn(data.slice(offset, offset + this.encodedLength));

    if (bytes.isZero()) {
      return [false, offset + this.encodedLength];
    }

    if (!bytes.eq(bn(1))) {
      throw new FuelError(ErrorCode.DECODE_ERROR, `Invalid boolean value.`);
    }

    return [true, offset + this.encodedLength];
  }
}
