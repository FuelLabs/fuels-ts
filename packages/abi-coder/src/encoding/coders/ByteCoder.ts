import { ErrorCode, FuelError } from '@fuel-ts/errors';
import { bn } from '@fuel-ts/math';

import { WORD_SIZE } from '../../utils/constants';

import { Coder } from './AbstractCoder';
import { BigNumberCoder } from './BigNumberCoder';

export class ByteCoder extends Coder<number[], Uint8Array> {
  static memorySize = 1;
  constructor() {
    super('struct', 'struct Bytes', WORD_SIZE);
  }

  encode(value: number[] | Uint8Array): Uint8Array {
    const bytes = value instanceof Uint8Array ? value : new Uint8Array(value);
    const lengthBytes = new BigNumberCoder('u64').encode(bytes.length);

    return new Uint8Array([...lengthBytes, ...bytes]);
  }

  decode(data: Uint8Array, offset: number): [Uint8Array, number] {
    if (data.length < WORD_SIZE) {
      throw new FuelError(ErrorCode.DECODE_ERROR, `Invalid byte data size.`);
    }

    const offsetAndLength = offset + WORD_SIZE;
    const lengthBytes = data.slice(offset, offsetAndLength);
    const length = bn(new BigNumberCoder('u64').decode(lengthBytes, 0)[0]).toNumber();

    const dataBytes = data.slice(offsetAndLength, offsetAndLength + length);

    if (dataBytes.length !== length) {
      throw new FuelError(ErrorCode.DECODE_ERROR, `Invalid bytes byte data size.`);
    }

    return [dataBytes, offsetAndLength + length];
  }
}
