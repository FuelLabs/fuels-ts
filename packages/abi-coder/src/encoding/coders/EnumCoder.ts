import { ErrorCode, FuelError } from '@fuel-ts/errors';
import { toNumber } from '@fuel-ts/math';
import { concat } from '@fuel-ts/utils';
import type { RequireExactlyOne } from 'type-fest';

import { OPTION_CODER_TYPE, VOID_TYPE } from '../../utils/constants';
import { hasNestedOption } from '../../utils/utilities';

import type { TypesOfCoder } from './AbstractCoder';
import { Coder } from './AbstractCoder';
import { BigNumberCoder } from './BigNumberCoder';

export type InputValueOf<TCoders extends Record<string, Coder>> = RequireExactlyOne<{
  [P in keyof TCoders]: TypesOfCoder<TCoders[P]>['Input'];
}>;
export type DecodedValueOf<TCoders extends Record<string, Coder>> = RequireExactlyOne<{
  [P in keyof TCoders]: TypesOfCoder<TCoders[P]>['Decoded'];
}>;

export class EnumCoder<TCoders extends Record<string, Coder>> extends Coder<
  InputValueOf<TCoders>,
  DecodedValueOf<TCoders>
> {
  name: string;
  coders: TCoders;
  #caseIndexCoder: BigNumberCoder;
  #encodedValueSize: number;
  #shouldValidateLength: boolean;

  constructor(name: string, coders: TCoders) {
    const caseIndexCoder = new BigNumberCoder('u64');
    const encodedValueSize = Object.values(coders).reduce(
      (min, coder) => Math.min(min, coder.encodedLength),
      0
    );
    super(`enum ${name}`, `enum ${name}`, caseIndexCoder.encodedLength + encodedValueSize);
    this.name = name;
    this.coders = coders;
    this.#caseIndexCoder = caseIndexCoder;
    this.#encodedValueSize = encodedValueSize;
    this.#shouldValidateLength = !(this.type === OPTION_CODER_TYPE || hasNestedOption(coders));
  }

  // Checks that we're handling a native enum that is of type void.
  #isNativeEnum(coder: Coder): boolean {
    return this.type !== OPTION_CODER_TYPE && coder.type === VOID_TYPE;
  }

  #encodeNativeEnum(value: string): Uint8Array {
    const valueCoder = this.coders[value];
    const encodedValue = valueCoder.encode([]);
    const caseIndex = Object.keys(this.coders).indexOf(value);

    const padding = new Uint8Array(this.#encodedValueSize - valueCoder.encodedLength);
    return concat([this.#caseIndexCoder.encode(caseIndex), padding, encodedValue]);
  }

  encode(value: InputValueOf<TCoders>): Uint8Array {
    if (typeof value === 'string' && this.coders[value]) {
      return this.#encodeNativeEnum(value);
    }

    const [caseKey, ...empty] = Object.keys(value);
    if (!caseKey) {
      throw new FuelError(ErrorCode.INVALID_DECODE_VALUE, 'A field for the case must be provided.');
    }
    if (empty.length !== 0) {
      throw new FuelError(ErrorCode.INVALID_DECODE_VALUE, 'Only one field must be provided.');
    }
    const valueCoder = this.coders[caseKey];
    const caseIndex = Object.keys(this.coders).indexOf(caseKey);
    if (caseIndex === -1) {
      const validCases = Object.keys(this.coders)
        .map((v) => `'${v}'`)
        .join(', ');
      throw new FuelError(
        ErrorCode.INVALID_DECODE_VALUE,
        `Invalid case '${caseKey}'. Valid cases: ${validCases}.`
      );
    }

    const encodedValue = valueCoder.encode(value[caseKey]);

    return new Uint8Array([...this.#caseIndexCoder.encode(caseIndex), ...encodedValue]);
  }

  #decodeNativeEnum(caseKey: string, newOffset: number): [DecodedValueOf<TCoders>, number] {
    return [caseKey as unknown as DecodedValueOf<TCoders>, newOffset];
  }

  decode(data: Uint8Array, offset: number): [DecodedValueOf<TCoders>, number] {
    if (this.#shouldValidateLength && data.length < this.encodedLength) {
      throw new FuelError(ErrorCode.DECODE_ERROR, `Invalid enum data size.`);
    }

    const caseBytes = new BigNumberCoder('u64').decode(data, offset)[0];
    const caseIndex = toNumber(caseBytes);
    const caseKey = Object.keys(this.coders)[caseIndex];
    if (!caseKey) {
      throw new FuelError(
        ErrorCode.INVALID_DECODE_VALUE,
        `Invalid caseIndex "${caseIndex}". Valid cases: ${Object.keys(this.coders)}.`
      );
    }
    const valueCoder = this.coders[caseKey];
    const offsetAndCase = offset + this.#caseIndexCoder.encodedLength;

    if (this.#shouldValidateLength && data.length < offsetAndCase + valueCoder.encodedLength) {
      throw new FuelError(ErrorCode.DECODE_ERROR, `Invalid enum data size.`);
    }

    const [decoded, newOffset] = valueCoder.decode(data, offsetAndCase);

    if (this.#isNativeEnum(this.coders[caseKey])) {
      return this.#decodeNativeEnum(caseKey, newOffset);
    }

    return [{ [caseKey]: decoded } as DecodedValueOf<TCoders>, newOffset];
  }
}
