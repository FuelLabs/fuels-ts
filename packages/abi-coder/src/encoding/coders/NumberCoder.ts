import { ErrorCode, FuelError } from '@fuel-ts/errors';
import { toNumber, toBytes } from '@fuel-ts/math';

import type { EncodingOptions } from '../../types/EncodingOptions';
import { WORD_SIZE } from '../../utils/constants';

import { Coder } from './AbstractCoder';

type NumberCoderType = 'u8' | 'u16' | 'u32';

const getLength = (baseType: NumberCoderType): number => {
  switch (baseType) {
    case 'u8':
      return 1;
    case 'u16':
      return 2;
    case 'u32':
      return 4;
    default:
      throw new FuelError(ErrorCode.TYPE_NOT_SUPPORTED, `Invalid number type: ${baseType}`);
  }
};

export class NumberCoder extends Coder<number, number> {
  baseType: NumberCoderType;
  options: EncodingOptions;

  constructor(
    baseType: NumberCoderType,
    options: EncodingOptions = {
      padToWordSize: false,
    }
  ) {
    const length = options.padToWordSize ? WORD_SIZE : getLength(baseType);
    super('number', baseType, length);
    this.baseType = baseType;
    this.options = options;
  }

  encode(value: number | string): Uint8Array {
    let bytes;

    try {
      bytes = toBytes(value);
    } catch (error) {
      throw new FuelError(ErrorCode.ENCODE_ERROR, `Invalid ${this.baseType}.`);
    }

    if (bytes.length > this.encodedLength) {
      throw new FuelError(ErrorCode.ENCODE_ERROR, `Invalid ${this.baseType}, too many bytes.`);
    }

    return toBytes(bytes, this.encodedLength);
  }

  decode(data: Uint8Array, offset: number): [number, number] {
    if (data.length < this.encodedLength) {
      throw new FuelError(ErrorCode.DECODE_ERROR, `Invalid number data size.`);
    }

    const bytes = data.slice(offset, offset + this.encodedLength);

    if (bytes.length !== this.encodedLength) {
      throw new FuelError(ErrorCode.DECODE_ERROR, `Invalid number byte data size.`);
    }

    return [toNumber(bytes), offset + this.encodedLength];
  }
}
