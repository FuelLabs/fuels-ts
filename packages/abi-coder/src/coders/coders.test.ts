/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { hexlify } from '@ethersproject/bytes';
import { bn, toHex } from '@fuel-ts/math';

import type Coder from './abstract-coder';
import ArrayCoder from './array';
import B256Coder from './b256';
import B512Coder from './b512';
import BooleanCoder from './boolean';
import ByteCoder from './byte';
import EnumCoder from './enum';
import NumberCoder from './number';
import StringCoder from './string';
import StructCoder from './struct';
import TupleCoder from './tuple';
import U64Coder from './u64';

const B256_ZERO = '0x0000000000000000000000000000000000000000000000000000000000000000';
const B256 = '0xd5579c46dfcc7f18207013e65b44e4cb4e2c2298f4ac457ba8f82743f31e930b';
const B512_ZERO =
  '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000';
const B512 =
  '0x8e9dda6f7793745ac5aacf9e907cae30b2a01fdf0d23b7750a85c6a44fca0c29f0906f9d1f1e92e6a1fb3c3dcef3cc3b3cdbaae27e47b9d9a4c6a4fce4cf16b2';
const U8_MAX = 2 ** 8 - 1;
const U16_MAX = 2 ** 16 - 1;
const U32_MAX = 2 ** 32 - 1;
const U64_MAX = bn(2).pow(64).sub(1);

/**
 * Tests for implementations of Coder.
 *
 * This is an array of tuples containing one item for each Coder.
 *
 * Each item is another tuple, containing:
 * - The name of the Coder
 * - A list of good cases we expect to succeed
 * - A list of bad cases we expect to fail
 *
 * And each case is also a tuple, containing:
 * - An instance of the Coder
 * - InputValue to provide to `coder.encode`
 * - DecodedValue to expect from `coder.decode`
 */
const testCases = [
  [
    'ArrayCoder',
    [
      [new ArrayCoder(new NumberCoder('u8'), 0), [], []],
      [new ArrayCoder(new NumberCoder('u8'), 4), [0, 13, 37, U8_MAX], [0, 13, 37, U8_MAX]],
    ],
    [
      // Under
      [new ArrayCoder(new NumberCoder('u8'), 1), [], []],
      // Over
      [new ArrayCoder(new NumberCoder('u8'), 1), [1, 2], [1, 2]],
      // Wrong
      [new ArrayCoder(new NumberCoder('u8'), 1), ['whoops'], ['whoops']],
      [new ArrayCoder(new NumberCoder('u8'), 1), [U8_MAX + 1], [U8_MAX + 1]],
    ],
  ],
  [
    'B256Coder',
    [
      [new B256Coder(), B256_ZERO, B256_ZERO],
      [new B256Coder(), B256, B256],
    ],
    [
      // Under
      [
        new B256Coder(),
        B256_ZERO.slice(0, B256_ZERO.length - 1),
        B256_ZERO.slice(0, B256_ZERO.length - 1),
      ],
      // Over
      [new B256Coder(), `${B256_ZERO}0`, `${B256_ZERO}0`],
      [new B256Coder(), `${B256_ZERO}00`, `${B256_ZERO}00`],
      // Wrong
      [new B256Coder(), 'whoops', 'whoops'],
    ],
  ],
  [
    'B512Coder',
    [
      [new B512Coder(), B512_ZERO, B512_ZERO],
      [new B512Coder(), B512, B512],
    ],
    [
      // Under
      [
        new B512Coder(),
        B512_ZERO.slice(0, B512_ZERO.length - 1),
        B512_ZERO.slice(0, B512_ZERO.length - 1),
      ],
      // Over
      [new B512Coder(), `${B512_ZERO}0`, `${B512_ZERO}0`],
      [new B512Coder(), `${B512_ZERO}00`, `${B512_ZERO}00`],
      // Wrong
      [new B512Coder(), 'whoops', 'whoops'],
    ],
  ],
  [
    'BooleanCoder',
    [
      [new BooleanCoder(), true, true],
      [new BooleanCoder(), false, false],
    ],
    [
      // Under
      [new BooleanCoder(), -1, -1],
      // Over
      [new BooleanCoder(), 2, 2],
      // Wrong
      [new BooleanCoder(), 'a', 'a'],
    ],
  ],
  [
    'ByteCoder',
    [
      // `string` inputs
      [new ByteCoder(), bn(0), 0],
      [new ByteCoder(), bn(U8_MAX), U8_MAX],
      // `number` inputs
      [new ByteCoder(), 0, 0],
      [new ByteCoder(), U8_MAX, 255],
    ],
    [
      // Under
      [new ByteCoder(), -1, -1],
      // Over
      [new ByteCoder(), bn(U8_MAX + 1), U8_MAX + 1],
      [new ByteCoder(), U8_MAX + 1, U8_MAX + 1],
      // Wrong
      [new ByteCoder(), 'whoops', 'whoops'],
    ],
  ],
  [
    'EnumCoder',
    [
      [
        new EnumCoder('TestEnum', { a: new BooleanCoder(), b: new U64Coder() }),
        { a: true },
        { a: true },
      ],
      [
        new EnumCoder('TestEnum', { a: new BooleanCoder(), b: new U64Coder() }),
        { b: bn(1337) },
        { b: bn(1337) },
      ],
      [
        new ArrayCoder(
          new EnumCoder('TestEnum', {
            a: new U64Coder(),
            b: new TupleCoder([new U64Coder(), new U64Coder()]),
          }),
          4
        ),
        [
          { a: bn(1337) },
          { b: [bn(1337), bn(1337)] },
          { a: bn(1337) },
          { b: [bn(1337), bn(1337)] },
        ],
        [
          { a: bn(1337) },
          { b: [bn(1337), bn(1337)] },
          { a: bn(1337) },
          { b: [bn(1337), bn(1337)] },
        ],
      ],
    ],
    [
      // Under
      [new EnumCoder('TestEnum', { a: new BooleanCoder(), b: new U64Coder() }), {}, {}],
      // Over
      [
        new EnumCoder('TestEnum', { a: new BooleanCoder(), b: new U64Coder() }),
        { a: true, b: bn(1337) },
        { a: true, b: bn(1337) },
      ],
      // Wrong
      [new EnumCoder('TestEnum', {}), {}, {}],
      [
        new EnumCoder('TestEnum', { a: new BooleanCoder(), b: new U64Coder() }),
        { b: true },
        { b: true },
      ],
    ],
  ],
  [
    'NumberCoder',
    [
      // `number` inputs
      [new NumberCoder('u8'), 0, 0],
      [new NumberCoder('u8'), U8_MAX, U8_MAX],
      [new NumberCoder('u16'), 0, 0],
      [new NumberCoder('u16'), U16_MAX, U16_MAX],
      [new NumberCoder('u32'), 0, 0],
      [new NumberCoder('u32'), U32_MAX, U32_MAX],

      // `u64` BigNumberish inputs
      [new U64Coder(), 0, bn(0)],
      [new U64Coder(), toHex(100), bn(100)],
      [new U64Coder(), bn(U8_MAX), bn(U8_MAX)],
      [new U64Coder(), U16_MAX, bn(U16_MAX)],
      [new U64Coder(), U32_MAX, bn(U32_MAX)],
      [new U64Coder(), U64_MAX, U64_MAX],
    ],
    [
      // Under
      [new NumberCoder('u8'), -1, -1],
      [new NumberCoder('u16'), -1, -1],
      [new NumberCoder('u32'), -1, -1],
      [new U64Coder(), -1, -1],
      // Over
      [new NumberCoder('u8'), U8_MAX + 1, U8_MAX + 1],
      [new NumberCoder('u16'), U16_MAX + 1, U16_MAX + 1],
      [new NumberCoder('u32'), U32_MAX + 1, U32_MAX + 1],
      [new U64Coder(), bn(U64_MAX.add(1)), bn(U64_MAX.add(1))],
      // Wrong
      [new NumberCoder('u8'), 'whoops', 'whoops'],
    ],
  ],
  [
    'StringCoder',
    [
      [new StringCoder(0), '', ''],
      [new StringCoder(U8_MAX), 'a'.repeat(U8_MAX), 'a'.repeat(U8_MAX)],
    ],
    [
      // Under
      [new StringCoder(0), 'a', 'a'],
      [new StringCoder(1), '', ''],
      // Over
      [new StringCoder(1), 'aa', 'aa'],
      // Wrong
      [new StringCoder(1), 'aa', 'aa'],
    ],
  ],
  [
    'StructCoder',
    [
      [new StructCoder('TestStruct', {}), {}, {}],
      [
        new StructCoder('TestStruct', { a: new BooleanCoder(), b: new U64Coder() }),
        { a: true, b: 1337 },
        { a: true, b: bn(1337) },
      ],
      [
        new StructCoder('TestStruct', { a: new BooleanCoder(), b: new U64Coder() }),
        { a: true, b: bn(1337) },
        { a: true, b: bn(1337) },
      ],
    ],
    [
      // Under
      [
        new StructCoder('TestStruct', { a: new BooleanCoder(), b: new U64Coder() }),
        { a: true },
        { a: true },
      ],
      [new StructCoder('TestStruct', { a: new BooleanCoder() }), {}, {}],
      // Over
      [
        new StructCoder('TestStruct', { a: new BooleanCoder() }),
        { a: true, b: 1337 },
        { a: true, b: 1337 },
      ],
      [new StructCoder('TestStruct', {}), { a: true }, { a: true }],
      // Wrong
      [new StructCoder('TestStruct', { b: new U64Coder() }), { b: U8_MAX + 1 }, { b: U8_MAX + 1 }],
      [new StructCoder('TestStruct', { b: new U64Coder() }), { b: 'whoops' }, { b: 'whoops' }],
    ],
  ],
  [
    'TupleCoder',
    [
      [new TupleCoder([]), [], []],
      [new TupleCoder([new U64Coder(), new U64Coder()]), [13, 37], [bn(13), bn(37)]],
    ],
    [
      // Under
      [new TupleCoder([new NumberCoder('u8')]), [], []],
      // Over
      [new TupleCoder([new NumberCoder('u8')]), [U8_MAX, U8_MAX], [U8_MAX, U8_MAX]],
      // Wrong
      [new TupleCoder([new NumberCoder('u8')]), [U8_MAX + 1], [U8_MAX + 1]],
      [new TupleCoder([new NumberCoder('u8')]), ['whoops'], ['whoops']],
    ],
  ],
] as const;

describe.each(testCases)('%s', (coderName, goodCases, badCases) => {
  it.each(
    goodCases.map(([coder, input, output]): [string, any, any, Coder] => [
      coder.type,
      input,
      output,
      coder,
    ])
  )('as a %s can encode %p then decode %p', (abiType, input, output, coder) => {
    const encoded = coder.encode(input);
    expect(hexlify(encoded)).toMatchSnapshot();
    const [decoded, length] = coder.decode(encoded, 0);
    expect(length).toEqual(encoded.length);

    expect(JSON.stringify(decoded)).toEqual(JSON.stringify(output));
  });
  it.each(
    badCases.map(([coder, input, output]): [string, any, any, Coder] => [
      coder.type,
      input,
      output,
      coder,
    ])
  )('as a %s can not encode %p then decode %p', (abiType, input, output, coder) => {
    expect(() => {
      const encoded = coder.encode(input);
      const [decoded] = coder.decode(encoded, 0);
      expect(decoded).toEqual(output);
    }).toThrow();
  });
});

describe('ArrayCoder', () => {
  it('will throw when value to encode is not array', () => {
    const coder = new ArrayCoder(new NumberCoder('u8'), 1);
    const nonArrayInput = { ...[1] };
    expect(() => {
      coder.encode(nonArrayInput);
    }).toThrow('expected array value');
  });
});

describe('B256Coder', () => {
  it('will throw when the decoded value length is not equal to 32 bytes', () => {
    const coder = new B256Coder();
    const input = new Uint8Array(Array.from(Array(32).keys()));
    expect(() => {
      coder.decode(input, 1);
    }).toThrow('Invalid size for b256');
  });
});

describe('B512Coder', () => {
  it('will throw when the decoded value length is not equal to 64 bytes', () => {
    const coder = new B512Coder();
    const input = new Uint8Array(Array.from(Array(64).keys()));
    expect(() => {
      coder.decode(input, 1);
    }).toThrow('Invalid size for b512');
  });
});

describe('BooleanCoder', () => {
  it('will throw when the decoded value is not a valid boolean value', () => {
    const coder = new BooleanCoder();
    const input = new Uint8Array(Array.from(Array(4).keys()));
    expect(() => {
      coder.decode(input, 1);
    }).toThrow('Invalid boolean value');
  });
});

describe('ByteCoder', () => {
  it('will throw when the decoded value length is invalid', () => {
    const coder = new ByteCoder();
    const input = new Uint8Array(Array.from(Array(4).keys()));
    expect(() => {
      coder.decode(input, 1);
    }).toThrow('Invalid Byte');
  });
});

describe('EnumCoder', () => {
  it('is typed correctly', () => {
    const coder = new EnumCoder('TestEnum', { a: new BooleanCoder(), b: new U64Coder() });

    // Good
    expect(() => coder.encode({ a: true })).not.toThrow();
    expect(() => coder.encode({ b: bn(1337) })).not.toThrow();
    // Under
    expect(() =>
      coder.encode(
        // @ts-expect-error
        {}
      )
    ).toThrow();
    // Over
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { a: true, b: 1337 }
      )
    ).toThrow();
    // Wrong
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { a: 1337 }
      )
    ).not.toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { b: true }
      )
    ).toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { uwu: 42 }
      )
    ).toThrow();
  });
  it('will throw when decoded value accesses an invalid coder index', () => {
    const coder = new EnumCoder('TestEnum', { a: new BooleanCoder(), b: new U64Coder() });
    const input = new Uint8Array(Array.from(Array(3).keys()));
    expect(() => {
      coder.decode(input, 1);
    }).toThrow('Invalid caseIndex');
  });
});

describe('StructCoder', () => {
  it('is typed correctly', () => {
    const coder = new StructCoder('TestStruct', {
      a: new BooleanCoder(),
      b: new U64Coder(),
    });

    // Good
    expect(() => coder.encode({ a: true, b: bn(1337) })).not.toThrow();
    // Under
    expect(() =>
      coder.encode(
        // @ts-expect-error
        {}
      )
    ).toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { a: true }
      )
    ).toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { b: bn(1337) }
      )
    ).toThrow();
    // Over
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { a: true, b: bn(1337), c: false }
      )
    ).not.toThrow();
    // Wrong
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { a: 1337 }
      )
    ).toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { b: true }
      )
    ).toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { uwu: 42 }
      )
    ).toThrow();
  });
});

describe('TupleCoder', () => {
  it('is typed correctly', () => {
    const coder = new TupleCoder<[BooleanCoder, U64Coder]>([new BooleanCoder(), new U64Coder()]);

    // Good
    expect(() => coder.encode([true, bn(1337)])).not.toThrow();
    // Under
    expect(() =>
      coder.encode(
        // @ts-expect-error
        []
      )
    ).toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        [true]
      )
    ).toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        [bn(1337)]
      )
    ).toThrow();
    // Over
    expect(() =>
      coder.encode(
        // @ts-expect-error
        [true, bn(1337), false]
      )
    ).toThrow();
    // Wrong
    expect(() =>
      coder.encode(
        // @ts-expect-error
        [bn(1337), true]
      )
    ).toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        ['true', bn(1337)]
      )
    ).not.toThrow();
    expect(() =>
      coder.encode(
        // @ts-expect-error
        { uwu: 42 }
      )
    ).toThrow();
  });
});
