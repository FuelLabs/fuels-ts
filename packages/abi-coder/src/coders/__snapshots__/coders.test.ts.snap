// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayCoder as a [u8; 0] can encode [] then decode [] 1`] = `"0x"`;

exports[`ArrayCoder as a [u8; 4] can encode [0, 13, 37, 255] then decode [0, 13, 37, 255] 1`] = `"0x0000000000000000000000000000000d000000000000002500000000000000ff"`;

exports[`B256Coder as a b256 can encode "0x0000000000000000000000000000000000000000000000000000000000000000" then decode "0x0000000000000000000000000000000000000000000000000000000000000000" 1`] = `"0x0000000000000000000000000000000000000000000000000000000000000000"`;

exports[`B256Coder as a b256 can encode "0xd5579c46dfcc7f18207013e65b44e4cb4e2c2298f4ac457ba8f82743f31e930b" then decode "0xd5579c46dfcc7f18207013e65b44e4cb4e2c2298f4ac457ba8f82743f31e930b" 1`] = `"0xd5579c46dfcc7f18207013e65b44e4cb4e2c2298f4ac457ba8f82743f31e930b"`;

exports[`B512Coder as a b512 can encode "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000" then decode "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000" 1`] = `"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`B512Coder as a b512 can encode "0x8e9dda6f7793745ac5aacf9e907cae30b2a01fdf0d23b7750a85c6a44fca0c29f0906f9d1f1e92e6a1fb3c3dcef3cc3b3cdbaae27e47b9d9a4c6a4fce4cf16b2" then decode "0x8e9dda6f7793745ac5aacf9e907cae30b2a01fdf0d23b7750a85c6a44fca0c29f0906f9d1f1e92e6a1fb3c3dcef3cc3b3cdbaae27e47b9d9a4c6a4fce4cf16b2" 1`] = `"0x8e9dda6f7793745ac5aacf9e907cae30b2a01fdf0d23b7750a85c6a44fca0c29f0906f9d1f1e92e6a1fb3c3dcef3cc3b3cdbaae27e47b9d9a4c6a4fce4cf16b2"`;

exports[`BooleanCoder as a boolean can encode false then decode false 1`] = `"0x0000000000000000"`;

exports[`BooleanCoder as a boolean can encode true then decode true 1`] = `"0x0000000000000001"`;

exports[`ByteCoder as a byte can encode "0x0" then decode 0 1`] = `"0x0000000000000000"`;

exports[`ByteCoder as a byte can encode "0xff" then decode 255 1`] = `"0x00000000000000ff"`;

exports[`ByteCoder as a byte can encode 0 then decode 0 1`] = `"0x0000000000000000"`;

exports[`ByteCoder as a byte can encode 255 then decode 255 1`] = `"0x00000000000000ff"`;

exports[`EnumCoder as a [enum TestEnum; 4] can encode [[Object], [Object], [Object], [Object]] then decode [[Object], [Object], [Object], [Object]] 1`] = `"0x000000000000000000000000000000000000000000000539000000000000000100000000000005390000000000000539000000000000000000000000000000000000000000000539000000000000000100000000000005390000000000000539"`;

exports[`EnumCoder as a enum TestEnum can encode {"a": true} then decode {"a": true} 1`] = `"0x00000000000000000000000000000001"`;

exports[`EnumCoder as a enum TestEnum can encode {"b": [BN]} then decode {"b": [BN]} 1`] = `"0x00000000000000010000000000000539"`;

exports[`NumberCoder as a u8 can encode 0 then decode 0 1`] = `"0x0000000000000000"`;

exports[`NumberCoder as a u8 can encode 255 then decode 255 1`] = `"0x00000000000000ff"`;

exports[`NumberCoder as a u16 can encode 0 then decode 0 1`] = `"0x0000000000000000"`;

exports[`NumberCoder as a u16 can encode 65535 then decode 65535 1`] = `"0x000000000000ffff"`;

exports[`NumberCoder as a u32 can encode 0 then decode 0 1`] = `"0x0000000000000000"`;

exports[`NumberCoder as a u32 can encode 4294967295 then decode 4294967295 1`] = `"0x00000000ffffffff"`;

exports[`NumberCoder as a u64 can encode "0x64" then decode "0x64" 1`] = `"0x0000000000000064"`;

exports[`NumberCoder as a u64 can encode "0xff" then decode "0xff" 1`] = `"0x00000000000000ff"`;

exports[`NumberCoder as a u64 can encode "0xffffffffffffffff" then decode "0xffffffffffffffff" 1`] = `"0xffffffffffffffff"`;

exports[`NumberCoder as a u64 can encode 0 then decode "0x0" 1`] = `"0x0000000000000000"`;

exports[`NumberCoder as a u64 can encode 65535 then decode "0xffff" 1`] = `"0x000000000000ffff"`;

exports[`NumberCoder as a u64 can encode 4294967295 then decode "0xffffffff" 1`] = `"0x00000000ffffffff"`;

exports[`StringCoder as a str[0] can encode "" then decode "" 1`] = `"0x"`;

exports[`StringCoder as a str[255] can encode "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" then decode "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" 1`] = `"0x61616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616161616100"`;

exports[`StructCoder as a struct TestStruct can encode {"a": true, "b": [BN]} then decode {"a": true, "b": [BN]} 1`] = `"0x00000000000000010000000000000539"`;

exports[`StructCoder as a struct TestStruct can encode {"a": true, "b": 1337} then decode {"a": true, "b": [BN]} 1`] = `"0x00000000000000010000000000000539"`;

exports[`StructCoder as a struct TestStruct can encode {} then decode {} 1`] = `"0x"`;

exports[`TupleCoder as a () can encode [] then decode [] 1`] = `"0x"`;

exports[`TupleCoder as a (u64, u64) can encode [13, 37] then decode [[BN], [BN]] 1`] = `"0x000000000000000d0000000000000025"`;
